mender-auth/http_forwarder/http_forwarder.cpp

```diff
@@ -0,0 +1,258 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <mender-auth/http_forwarder.hpp>
+
+#include <random>
+
+namespace mender {
+namespace auth {
+namespace http_forwarder {
+
+ForwardObject::ForwardObject(const http::ClientConfig &config, events::EventLoop &event_loop) :
+	client_(config, event_loop),
+	logger_("http_forwarder") {
+}
+
+Server::Server(
+	const http::ServerConfig &server_config,
+	const http::ClientConfig &client_config,
+	events::EventLoop &loop) :
+	logger_("http_forwarder"),
+	event_loop_ {loop},
+	server_ {server_config, loop},
+	cancelled_ {make_shared<bool>(true)},
+	client_config_ {client_config} {
+}
+
+Server::~Server() {
+	*cancelled_ = true;
+	cancelled_ = make_shared<bool>(true);
+	Cancel();
+}
+
+void Server::Cancel() {
+	*cancelled_ = true;
+	cancelled_ = make_shared<bool>(true);
+	connections_.clear();
+	server_.Cancel();
+}
+
+error::Error Server::AsyncForward(const string &listen_url, const string &target_url) {
+	if (!*cancelled_) {
+		return error::Error(
+			make_error_condition(errc::operation_in_progress),
+			"HTTP forwarding already in progress");
+	}
+
+	*cancelled_ = false;
+
+	http::BrokenDownUrl target_address;
+
+	// We don't actually need target_address here, but break it down here anyway to avoid that
+	// the error shows up much later when the first connection is made.
+	auto err = http::BreakDownUrl(target_url, target_address);
+	if (err != error::NoError) {
+		return err.WithContext("HTTP forwarder: Invalid target address");
+	}
+	target_url_ = target_url;
+
+	auto cancelled = cancelled_;
```

Comment:
Because C++ doesn't allow direct capture of member variables. However, as indicated in another comment, I should make this a reference, which avoids one copy of two (still one copy in the lambda, which we want).